### ISSUE TITLE
Golang and multiple files

### ISSUE THREAD

**Issue by henrikbjorn** [neomake/neomake#134]
If you have two files `main.go` and `handler.go`. Where `main.go` uses a function declared in `handler.go` the marker will fail as it doesnt see that handler.go exists in the same package.

This was not a problem when using Syntastic, so i think it is something inside Neomake.

**Comment by renstrom** [comment]
You have to use `:Neomake!` for "project" wide checkers, the regular `:Neomake` only validates the current buffer.

**Comment by henrikbjorn** [comment]
Hmm that seems very slow compared to the syntastic checkers.

**Comment by Neki** [comment]
'go' is defined as a file maker for the 'go' filetype, it should work with `:Neomake` out of the box. If not, I think it can be considered as a bug with the checker. I don't know about Go so I can't really say more.

**Comment by henrikbjorn** [comment]
Have looked quickly at the syntastic code and it uses cwd as the package instead of figuring out what its name is. This works quite well, but i dont know how to make it use the same command.

https://github.com/scrooloose/syntastic/blob/master/syntax_checkers/go/go.vim#L79

 Also it seems like they do additional checks that could be useful aswell.

**Comment by Neki** [comment]
Neomake does not (currently) support changing the working directory before running a command, so you can not easily mirror the Syntastic configuration.

**Comment by renstrom** [comment]
Would it work to add a `cd` command to the `args` list?

**Comment by Neki** [comment]
The args list is placed after the executable name when building the command, and the 'exe' part must be an executable name (so 'exe': `cd folder; go build` won't do). I guess you could try a hack such as `args: ['; cd folder; go build'], exe: 'true'` but I feel dirty just suggesting it :)
What you could also do is creating a wrapper around the go compiler that accepts a folder as a parameter and changes the current directory before calling the actual go compiler.
But it is better to allow the change the cwd directly in Neomake. I am not sure about how Neovim job control handles cwd, but hopefully it will be an easy change.

**Comment by Neki** [comment]
I am working on adding the 'cwd' option to Neomake, should be straightforward. To fully reproduce the Syntastic configuration, we will need more than this, but it should solve the biggest pain point for Go users.

**Comment by peter-edge** [comment]
Hey @Neki any update on this? Happy to help if you point me in the right direction, I really want to switch over to neomake :)

**Comment by Neki** [comment]
Sorry I let this issue aside for so long (especially for such a small patch), I pushed the relevant PR at #158.
I don't use Go so I can not reliably update the Go maker, but once #158 is merged I think this patch should do the trick:

```patch
diff --git a/autoload/neomake/makers/ft/go.vim b/autoload/neomake/makers/ft/go.vim
index 4fe364b..516e32c 100644
--- a/autoload/neomake/makers/ft/go.vim
+++ b/autoload/neomake/makers/ft/go.vim
@@ -15,7 +15,8 @@ function! neomake#makers#ft#go#go()
             \ '%E%f:%l:%c:%m,' .
             \ '%E%f:%l:%m,' .
             \ '%C%\s%\+%m,' .
-            \ '%-G#%.%#'
+            \ '%-G#%.%#',
+        \ 'cwd': '%:p:h'
         \ }
 endfunction
 ```

Let me know how it goes.

**Comment by peter-edge** [comment]
Ah very nice. Thanks. When it's merged, I'll give it a try :)

**Comment by Neki** [comment]
It's merged now :)

**Comment by peter-edge** [comment]
Perfect :)

**Comment by peter-edge** [comment]
Hmm, still doesn't work with multiple files, any ideas? See attached screenshot:

![image](https://cloud.githubusercontent.com/assets/4228796/8850212/7939b71e-3146-11e5-995b-c8fbfce80bd9.png)

**Comment by peter-edge** [comment]
If you want to test that for reproduction, that repository is public:

Assuming you are using a mac, install golang (if not, I can give you instructions for linux too):

```
brew install golang
export GOPATH=~/go
mkdir -p "${GOPATH}"
```

Then, to get this replicated:

```
go get github.com/pachyderm/pachyderm/...
cd "${GOPATH}/src/github.com/pachyderm/pachyderm"
make test-deps
vim src/pps/graph/graph.go
```

And save the file with `:w` obviously :)

**Comment by Neki** [comment]
Could you post the definition of the Go maker that you are using? (just to make sure things are clear, I did not update to Go maker myself).

**Comment by peter-edge** [comment]
I'm not sure what you are referring to, what do you mean by "go maker"
exactly? :)

**Comment by peter-edge** [comment]
Oh, make is just make :)

```
$ make --version
GNU Make 4.0
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2013 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by peter-edge** [comment]
On my mac, it's this:

```
$ make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

**Comment by Neki** [comment]
Sorry, I was not clear. I did not do anything that change the way Go is supported in Neomake ; I only added what was necessary for someone else to solve this issue.

The patch I posted above should do the trick, but I have not tested it. If nobody does it before me, I'll see If I can easily install Go and try the patch, but I make no promise about when I'll do that.

**Comment by peter-edge** [comment]
Ah ok, I'll try it locally, sorry for the confusion :)

**Comment by peter-edge** [comment]
@Neki i can verify that that patch does work, i just tested it, so if you added it, that would be great :)

**Comment by henrikbjorn** [comment]
@Neki really? I tried and it just stopped reporting anything for go.

**Comment by peter-edge** [comment]
@henrikbjorn no it still doesn't work actually, I was mistaken

**Comment by peter-edge** [comment]
@henrikbjorn it works for me with

1. the above patch
2. `au BufWritePost *.go Neomake!`

The exited 2 is because you're now calling go with the file name which you do not want

**Comment by henrikbjorn** [comment]
`Neomake!` worked before aswell as it runs on all the files, and takes a long time.

**Comment by torkelo** [comment]
anyone solved this?

**Comment by redlolgeerf** [comment]
@henrikbjorn , what is your pwd, when you invoke "Neomake!"?

As I get it, you should always use "Neomake!" command, not "Neomake", to invoke "go" maker, as "go build" works with the package, not file, and "Neomake" always adds current file name to args list.

**Comment by henrikbjorn** [comment]
@redlolgeerf i am not interested in doing the whole package but just the file, also go will load the package if the pwd is correct. Just as syntastic does.

**Comment by redlolgeerf** [comment]
What do you mean by "not doing the whole package but just the file"?
Syntastic does smth like "cd package_dir && go build . && cd -"

**Comment by redlolgeerf** [comment]
@Neki, your patch does its job for "Neomake" command, but not for "Neomake!". "Neomake!" uses "makepgr" maker, not "go", so it gets no "cwd" option.

**Comment by vito** [comment]
Here's a workaround I landed on, with the help of the `cwd` patch above:

```vim
let g:neomake_go_gobuild_maker = {
    \ 'exe': 'sh',
    \ 'args': ['-c', 'go build -o ' . neomake#utils#DevNull() . ' .'],
    \ 'errorformat':
        \ '%W%f:%l: warning: %m,' .
        \ '%E%f:%l:%c:%m,' .
        \ '%E%f:%l:%m,' .
        \ '%C%\s%\+%m,' .
        \ '%-G#%.%#',
    \ 'cwd': '%:p:h'
    \ }

let g:neomake_go_enabled_makers = ['gobuild', (...)]
```

The trick is to work around this logic: https://github.com/benekastah/neomake/blob/dc65a7a5d85670c84fc0055d19fa6901ae96ef93/autoload/neomake.vim#L70-L73

...wherein it automatically appends a filename if not already present. I tried a few variants before landing on this one, which just uses `sh` and ignores any arguments passed to it. So the command that ends up running is something like `sh -c 'go build -o /dev/null .' /path/to/file.go`.

I'm not sure if that's something other makers would need though. I don't want to/can't use `Neomake!` as I'd rather have it go through the various makers I've configured instead of just `makeprg` (which can only have one `errorformat`.

This probably only applies to makers for languages like Go where the full package is what defines things, and not individual files with explicit imports/etc.

**Comment by vito** [comment]
One caveat with the above approach: with `cwd` set, the file paths printed will be relative to the package itself, i.e. `foo.go` instead of `package/foo.go`. So if you try to go to the location via `:lnext` it'll actually open a new file. Hrmmm...

**Comment by henrikbjorn** [comment]
@vito where do i paste this, it doesnt seem like the other functions that is used for the makers.

**Comment by vito** [comment]
@henrikbjorn it goes in your vim config

**Comment by henrikbjorn** [comment]
Seems to workish. It does not trigger a scratch window telling me the errors.

**Comment by vito** [comment]
@henrikbjorn that's part of neomake's config, not my workaround; you want `let g:neomake_open_list = 2` (or some such).

**Comment by henrikbjorn** [comment]
@vito If i remove the workaround the bottom scratch window pops open :)

**Comment by nhooyr** [comment]
Any updates on this?

**Comment by blueyed** [comment]
There have been improvements to the Golang makers, and https://github.com/neomake/neomake/pull/1326 should make it work properly really now.
